apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableReadRowsRequestManager.cs

```diff
@@ -0,0 +1,187 @@
+                {
+                    return comparison;
+                }
+            }
+            return size1.CompareTo(size2);
+        }
+
+        // Memeber variables form the constructor.
+        private readonly ReadRowsRequest _originalRequest;
+
+        // The number of rows read so far.
+        private long _rowCount;
+
+        /// <summary>
+        /// Constructor for ResumingStreamingResultScanner.
+        /// </summary>
+        /// <param name="originalRequest">
+        /// Original ReadRowsRequest containg all of the parameters of the API call.
+        /// </param>
+        public BigtableReadRowsRequestManager(ReadRowsRequest originalRequest)
+        {
+            _originalRequest = originalRequest;
+        }
+
+        /// <summary>
+        /// Update last found Rowkey
+        /// </summary>
+        public ByteString LastFoundKey { get; set; }
+
+        /// <summary>
+        /// Update amount of rows read so far
+        /// </summary>
+        public long RowCount
+        {
+            get => _rowCount;
+            set => _rowCount += value;
+        }
+
+        /// <summary>
+        /// Builds and returns updated subrequest
+        /// </summary>
+        /// <returns></returns>
+        public ReadRowsRequest BuildUpdatedRequest()
+        {
+            ReadRowsRequest newReadRowsRequest = new ReadRowsRequest()
+
+            {
+                TableNameAsTableName = _originalRequest.TableNameAsTableName,
+            };
+
+            #region Set up RowFilter
+
+            if (_originalRequest.Filter != null)
+            {
+                newReadRowsRequest.Filter = new RowFilter();
+                newReadRowsRequest.Filter = _originalRequest.Filter;
+            }
+
+            #endregion
+
+            #region Set up RowsLimit
+
+            // If the row limit is set, update it.
+            long numRowsLimit = _originalRequest.RowsLimit;
+
+            // Update <see cref="numRowsLimit"/> by removing the number of rows already read.
+            if ((numRowsLimit -= RowCount) > 0)
+            // The remaining number of rows must be greater than 0.
+            {
+                newReadRowsRequest.RowsLimit = numRowsLimit;
+            }
+
+            #endregion
+
+            #region Set up RowSet
+
+            newReadRowsRequest.Rows = new RowSet();
+            newReadRowsRequest.Rows = FilterRows();
+
+            #endregion
+
+            return newReadRowsRequest;
+        }
+
+        private RowSet FilterRows()
+        {
+            RowSet originalRows = _originalRequest.Rows;
+
+            if (LastFoundKey == null)
+            {
+                return originalRows;
+            }
+
+            RowSet newRowSet = new RowSet();
+
+            foreach (ByteString key in originalRows.RowKeys)
+            {
+                if (!StartKeyIsAlreadyRead(key))
+                {
+                    newRowSet.RowKeys.Add(key);
+                }
+            }
```

Comment:
You can replace loop with:
```csharp
newRowSet.RowKeys.AddRange(originalRows.RowKeys.Select(key => !StartKeyIsAlreadyRead(key)))
```

Actually, if you make a negated version (`StartKeyIsntAlreadyRead`), it can be even simpler:
```csharp
newRowSet.RowKeys.AddRange(originalRows.RowKeys.Select(StartKeyIsntAlreadyRead))
```